diff --git a/arch/x86/boot/compressed/misc.c b/arch/x86/boot/compressed/misc.c
index 0f41ca0e5..04e9d4d38 100644
--- a/arch/x86/boot/compressed/misc.c
+++ b/arch/x86/boot/compressed/misc.c
@@ -115,6 +115,16 @@ static void serial_putchar(int ch)
 	outb(ch, early_serial_base + TXR);
 }
 
+static void ps5_putchar(uint8_t ch)
+{
+	volatile uint32_t* titania_uart = (volatile uint32_t*)0xc1010100;
+	if (ch == '\n')
+		ps5_putchar('\r');
+	while ((titania_uart[3] & 0x800))
+		cpu_relax();
+	titania_uart[1] = ch;
+}
+
 void __putstr(const char *s)
 {
 	int x, y, pos;
@@ -129,6 +139,12 @@ void __putstr(const char *s)
 		}
 	}
 
+	if (boot_params_ptr->hdr.hardware_subarch == X86_SUBARCH_PS5) {
+		const char *str = s;
+		while (*str)
+			ps5_putchar(*str++);
+	}
+
 	if (lines == 0 || cols == 0)
 		return;
 
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
index 492ce94c4..c25b073ca 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
@@ -43,7 +43,6 @@
 
 #ifdef CONFIG_X86_PS5
 #include <linux/ps5.h>
-#include <asm/setup.h>
 #endif
 
 #include "amdgpu.h"
@@ -2406,7 +2405,7 @@ static int amdgpu_pci_probe(struct pci_dev *pdev,
 	bool supports_atomic = false;
 
 #ifdef CONFIG_X86_PS5
-	if (boot_params.hdr.hardware_subarch == X86_SUBARCH_PS5_ && !isHdmiEnabled()) {
+	if (!isHdmiEnabled()) {
 		return -EPROBE_DEFER;
 	}
 #endif
diff --git a/drivers/tty/serial/Kconfig b/drivers/tty/serial/Kconfig
index f86775cfd..d097d7d72 100644
--- a/drivers/tty/serial/Kconfig
+++ b/drivers/tty/serial/Kconfig
@@ -1619,6 +1619,13 @@ config SERIAL_ESP32_ACM
 	  snippet may be used:
 	    earlycon=esp32s3acm,mmio32,0x60038000
 
+config SERIAL_TITANIA
+	tristate "Titania UART (as found in the PS5)"
+	depends on X86 || COMPILE_TEST
+	select SERIAL_EARLYCON
+	help
+	  Earlycon driver for the Titania UART, as found in the Sony PS5.
+
 endmenu
 
 config SERIAL_MCTRL_GPIO
diff --git a/drivers/tty/serial/Makefile b/drivers/tty/serial/Makefile
index a2ccbc508..6315f8072 100644
--- a/drivers/tty/serial/Makefile
+++ b/drivers/tty/serial/Makefile
@@ -89,6 +89,7 @@ obj-$(CONFIG_SERIAL_TEGRA)		+= serial-tegra.o
 obj-$(CONFIG_SERIAL_TEGRA_TCU)		+= tegra-tcu.o
 obj-$(CONFIG_SERIAL_TEGRA_UTC)		+= tegra-utc.o
 obj-$(CONFIG_SERIAL_TIMBERDALE)		+= timbuart.o
+obj-$(CONFIG_SERIAL_TITANIA)		+= titania.o
 obj-$(CONFIG_SERIAL_TXX9)		+= serial_txx9.o
 obj-$(CONFIG_SERIAL_UARTLITE)		+= uartlite.o
 obj-$(CONFIG_SERIAL_VT8500)		+= vt8500_serial.o
diff --git a/drivers/tty/serial/titania.c b/drivers/tty/serial/titania.c
new file mode 100644
index 000000000..b3fc4dcaa
--- /dev/null
+++ b/drivers/tty/serial/titania.c
@@ -0,0 +1,46 @@
+#include <linux/console.h>
+#include <linux/serial_core.h>
+#include <linux/platform_device.h>
+#include <linux/irqdomain.h>
+#include <linux/irq.h>
+#include <linux/tty_flip.h>
+#include <asm/irqdomain.h>
+#include <asm/hw_irq.h>
+#include <asm/io.h>
+#include <asm/setup.h>
+
+static void titania_early_putchar(void __iomem *registers, char c)
+{
+	while ((readl(registers+12) & 0x800));
+	writel(c, registers+4);
+}
+
+static void titania_early_write(struct console *con, const char *s, unsigned int n)
+{
+	void __iomem *registers = con->data;
+	for (unsigned int i = 0; i < n; ++i) {
+		if (s[i] == '\n')
+			titania_early_putchar(registers, '\r');
+		titania_early_putchar(registers, s[i]);
+	}
+}
+
+static int titania_early_read(struct console *con, char *s, unsigned int n)
+{
+	void __iomem *registers = (void __iomem*)con->data;
+	unsigned int i = 0;
+
+	while (i < n && (readl(registers+12) & 0x10))
+		s[i++] = readl(registers+0);
+
+	return i;
+}
+
+static int __init titania_early_console_setup(struct earlycon_device *device, const char *opt)
+{
+	device->con->data = (void*)early_ioremap(0xc1010100, 16);
+	device->con->write = titania_early_write;
+	device->con->read = titania_early_read;
+	return 0;
+}
+EARLYCON_DECLARE(titania, titania_early_console_setup);
